### PR TITLE
Expose RequestParams._meta in ClientSession.call_tool

### PR DESCRIPTION
**Pull request by samchenatti**
## Motivation and Context
This change allow clients to send metadata through the `_meta` field [as specified by the protocol](https://modelcontextprotocol.io/specification/2025-06-18/basic#meta).

It is important to expose `_meta` since there might be some metadata required by the server to control how the tool works (such as client preferences, its localization and so on). In some cases we also don't want the Agent/LLM to be aware of such data; it should be known by the MCP Host and Server only; therefore it cannot be exposed as tool usual arguments.

This feature has already been requested [by me](https://github.com/modelcontextprotocol/python-sdk/issues/1216) and [another user](https://github.com/modelcontextprotocol/python-sdk/issues/1197) - so it might benefit another people as well.

I also [started a discussion](https://github.com/jlowin/fastmcp/issues/1315) about `_meta` being exposed by FastMCP Client, but they will wait until its implemented in here.

## How Has This Been Tested?
I tested it on my local environment using a self-made MCP Server. 

Since `_meta` is currently being used for progress report tokens, I also ensured this feature doesn't break `progress_callback`.

I'd be happy to implement the required unit-tests if someone can point the best test-file for it.

## Breaking Changes
There are no breaking changes. The `python-sdk` already considers that `_meta` can carry data other than `progressToken`.

Also `types.RequestParams.Meta` model is already configured to allow extra key-values.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
I decided to expose `meta` as a plain dict instead of `types.RequestParams.Meta` since, from my understanding, `mcp.types` shouldn't be exposed to final users.

Metadata could be exposed by another interfaces as well, such as `list_tools`, `list_prompts` and etc. I'm sticking with `tool_call` for now, but would be happy to replicate it to the other interfaces.

### CODE REVIEW

File: src/mcp/client/session.py [hunk 1]
```diff
@@ -295,8 +296,7 @@ async def call_tool(
                 types.CallToolRequest(
                     method="tools/call",
                     params=types.CallToolRequestParams(
-                        name=name,
-                        arguments=arguments,
+                        name=name, arguments=arguments, _meta=types.RequestParams.Meta(**(meta or {}))
```
Comment (samchenatti) [inline src/mcp/client/session.py]: Handling meta separately would improve readability and avoid unnecessary call to PyDantic:
```python
        ...
        _meta: types.ReadResourceRequestParams.Meta | None = None
        if meta is not None:
            _meta = types.RequestParams.Meta(**meta)

        result = await self.send_request(
            types.ClientRequest(
                types.CallToolRequest(
                    method="tools/call",
                    params=types.CallToolRequestParams(name=name, arguments=arguments, _meta=_meta),
                )
            ),
```

File: src/mcp/client/session.py [hunk 2]
```diff
@@ -286,6 +286,7 @@ async def call_tool(
         name: str,
         arguments: dict[str, Any] | None = None,
         read_timeout_seconds: timedelta | None = None,
+        meta: dict[str, Any] | None = None,
```
Comment (dsp-ant) [inline src/mcp/client/session.py]: Note that this ordering breaks BC. If we go down this route we should probably just consistently 
```suggestion
        name: str,
        arguments: dict[str, Any] | None = None,
        *,
        read_timeout_seconds: timedelta | None = None,
        meta: dict[str, Any] | None = None,
```

Comment (samchenatti) [reply src/mcp/client/session.py]: Thanks for the review!

Since `read_timeout_seconds` is already part of the public interface as a positional argument, shouldn't it be:
```python
        name: str,
        arguments: dict[str, Any] | None = None,
        read_timeout_seconds: timedelta | None = None,
        *,
        meta: dict[str, Any] | None = None,
```
?

Comment (samchenatti) [reply src/mcp/client/session.py]: I'll follow with `meta` as the solely "forced" name parameter.

@dsp-ant I can make a new change if you want.

**Comment by danielabonfim** [comment]
This feature looks great and would be a valuable addition. Nice work! Would it be possible to get this approved for merge @dsp-ant

**Comment by samchenatti** [comment]
Hey, @dsp-ant!

Are there any other changes I need to make?

**Comment by discorev** [comment]
I'd also be interested in this, is there anything further that needs changing?

**Comment by samchenatti** [comment]
Hey, @dsp-ant, @ihrpr!

Is there anything else I need to fix in order to get this PR merged?

**Comment by aadarshsingh191198** [comment]
I'd also be interested in this! Requesting maintainers to merge it if there are no unwanted side effects of this change.

**Comment by maeryo** [comment]
Great work @samchenatti! This feature would be really valuable for the MCP ecosystem. I hope the conflicts can be resolved soon and this PR gets merged. The community clearly supports this addition based on all the positive feedback. Keep up the excellent work! 🚀

**Comment by samchenatti** [comment]
Thanks for the feedback, @maeryo and @felixweinberger!

I'll fix the conflicts by the end of the day

**Comment by samchenatti** [comment]
@felixweinberger, @maeryo conflicts solved!

**Comment by maeryo** [comment]
cool! I also checked today and noticed that the kotlin-sdk didn’t have this implementation yet, so I submitted a PR with a similar approach to yours. In addition to just adding the meta parameter, I paid extra attention to improving validation logic as well.

https://github.com/modelcontextprotocol/kotlin-sdk/pull/289

**Comment by samchenatti** [comment]
Hey, @maeryo, thanks for the feedback!

As suggested, I've implemented the validation here as well, but with a different approach.

Instead of validating _meta in the tool call only, I've implemented the validation directly within the types.RequestParams.Meta model. This seems more appropriate from a Pydantic standpoint.

I've also added unit tests with positive and negative keys, so you can confirm that all the rules are covered.

**Note**: Please be aware that these validations could be a breaking change, since users may already have non-conformant metadata keys in use.

**Edit:** I think it is a better idea to make a new PR focusing on the metadata key validation; see https://github.com/modelcontextprotocol/python-sdk/pull/1397

**Comment by samchenatti** [comment]
@felixweinberger @ihrpr let me know if there's anything else i need to change!

**Comment by Jaorow** [comment]
Hey @samchenatti
This is awesome and really helps out with something im trying to build! 

Do you have an ETA on when this might be in?

**Comment by samchenatti** [comment]
Hey @Jaorow!

I'm waiting on the maintainers’ feedback...

This feature has become even more important since OpenAI announced the App SDK.
They’re making heavy use of _meta to enable UI components.

By merging this PR, we’d allow other MCP Clients to do the same.

File: src/mcp/client/session.py [hunk 3]
```diff
@@ -273,16 +273,20 @@ async def call_tool(
...
+            _meta = types.RequestParams.Meta(**meta)
+
         result = await self.send_request(
             types.ClientRequest(
                 types.CallToolRequest(
-                    params=types.CallToolRequestParams(
-                        name=name,
-                        arguments=arguments,
-                    ),
+                    method="tools/call",
```
Comment (maxisbey) [inline src/mcp/client/session.py]: remove this, not needed as it's the default value

File: src/mcp/client/session.py [hunk 4]
```diff
@@ -273,16 +273,20 @@ async def call_tool(
         arguments: dict[str, Any] | None = None,
         read_timeout_seconds: timedelta | None = None,
         progress_callback: ProgressFnT | None = None,
+        *,
+        meta: dict[str, Any] | None = None,
     ) -> types.CallToolResult:
         """Send a tools/call request with optional progress callback support."""
 
+        _meta: types.ReadResourceRequestParams.Meta | None = None
```
Comment (maxisbey) [inline src/mcp/client/session.py]: Why is this ReadResourceRequestParams?

**Review by maxisbey — CHANGES_REQUESTED** [review]
I'd want to see some unit tests for this. You could add them in `tests/server/test_session.py`

**Comment by Jaorow** [comment]
> Hey @Jaorow!
> 
> I'm waiting on the maintainers’ feedback...
> 
> This feature has become even more important since OpenAI announced the App SDK. They’re making heavy use of _meta to enable UI components.
> 
> By merging this PR, we’d allow other MCP Clients to do the same.

@samchenatti Awesome, It will definitely open the door for more complex and structured functionality.

**Comment by samchenatti** [comment]
Hey, @maxisbey!

I made the requested changes and implemented the unit test.

~I had to cheat a bit in the test and manually set `_tool_output_schemas`~

After a good night’s sleep, I was able to properly implement the unit test (commit [2b61abba](https://github.com/samchenatti/mcp-python-sdk/commit/2b61abba1368a760a42149d084291b56bcd318f2)).

Now the client and server exchange all the necessary messages just like they would in a real scenario.

**Comment by samchenatti** [comment]
@maxisbey thanks for the approval!

@ihrpr, @dsp-ant can you guys also take a look on my changes?
